File: lib/src/crypto/SdkFilterInputStream.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.aliyun.oss.crypto;

import static com.aliyun.oss.crypto.SdkRuntime.shouldAbort;

import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;
import com.aliyun.oss.ClientErrorCode;
import com.aliyun.oss.ClientException;

/**
 * Base class for OSS Java SDK specific {@link FilterInputStream}.
 */
public class SdkFilterInputStream extends FilterInputStream {
    private volatile bool aborted = false;

    public SdkFilterInputStream(InputStream in) {
        super(in);
    }

    /**
     * @return The wrapped stream.
     */
    public InputStream getDelegateStream() {
        return in;
    }

    /**
     * Aborts the inputstream operation if thread is interrupted.
     * interrupted status of the thread is cleared by this method.
     * 
     * @throws ClientException with ClientErrorCode INPUTSTREAM_READING_ABORTED if thread aborted.
     */
    protected final void abortIfNeeded() {
        if (shouldAbort()) {
            abort();
            throw new ClientException("Thread aborted, inputStream aborted...",
                    ClientErrorCode.INPUTSTREAM_READING_ABORTED, null);
        }
    }

    public void abort() {
        if (in instanceof SdkFilterInputStream) {
            ((SdkFilterInputStream) in).abort();
        }
        aborted = true;
    }

    public bool isAborted() {
        return aborted;
    }

    @override
    public int read() throws IOException {
        abortIfNeeded();
        return in.read();
    }

    @override
    public int read(byte b[], int off, int len) throws IOException {
        abortIfNeeded();
        return in.read(b, off, len);
    }

    @override
    public long skip(long n) throws IOException {
        abortIfNeeded();
        return in.skip(n);
    }

    @override
    public int available() throws IOException {
        abortIfNeeded();
        return in.available();
    }

    @override
    public void close() throws IOException {
        in.close();
        abortIfNeeded();
    }

    @override
    public synchronized void mark(int readlimit) {
        abortIfNeeded();
        in.mark(readlimit);
    }

    @override
    public synchronized void reset() throws IOException {
        abortIfNeeded();
        in.reset();
    }

    @override
    public bool markSupported() {
        abortIfNeeded();
        return in.markSupported();
    }

    public void release() {
        if (in != null) {
            try {
                in.close();
            } catch (IOException ex) {
                // Ignore exception.
            }
        }
    }
}

File: lib/src/crypto/sdk_filter_input_stream.dart
import 'package:aliyun_oss_dart_sdk/src/client_error_code.dart';
import 'package:aliyun_oss_dart_sdk/src/client_exception.dart';
import 'package:aliyun_oss_dart_sdk/src/event/progress_input_stream.dart';

import 'sdk_runtime.dart';

/// Base class for OSS Java SDK specific {@link FilterInputStream}.
class SdkFilterInputStream extends FilterInputStream {
  bool aborted = false;

  SdkFilterInputStream(InputStream inputStream) : super(inputStream);

  /// Aborts the inputstream operation if thread is interrupted.
  /// interrupted status of the thread is cleared by this method.
  ///
  /// @throws ClientException with ClientErrorCode INPUTSTREAM_READING_ABORTED if thread aborted.
  void abortIfNeeded() {
    if (SdkRuntime.shouldAbort()) {
      abort();
      throw ClientException("Thread aborted, inputStream aborted...",
          ClientErrorCode.INPUTSTREAM_READING_ABORTED, null);
    }
  }

  void abort() {
    if (inputStream is SdkFilterInputStream) {
      (inputStream as SdkFilterInputStream).abort();
    }
    aborted = true;
  }

  bool isAborted() {
    return aborted;
  }

  @override
  int read([List<int>? list, int? off, int? len]) {
    abortIfNeeded();
    if (list == null) {
      return inputStream.read();
    }

    int offset = off ?? 0;
    int length = len ?? list.length;
    return inputStream.read(list, offset, length);
  }

  @override
  int skip(int n) {
    abortIfNeeded();
    return inputStream.skip(n);
  }

  @override
  int available() {
    abortIfNeeded();
    return inputStream.available();
  }

  @override
  void close() {
    inputStream.close();
    abortIfNeeded();
  }

  @override
  void mark(int readlimit) {
    abortIfNeeded();
    inputStream.mark(readlimit);
  }

  @override
  void reset() {
    abortIfNeeded();
    inputStream.reset();
  }

  @override
  bool markSupported() {
    abortIfNeeded();
    return inputStream.markSupported();
  }

  void release() {
    if (inputStream != null) {
      try {
        inputStream.close();
      } catch (ex) {
        // Ignore exception.
      }
    }
  }
}
